test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,446 @@
+	})
+
+	ginkgo.It("Should be able to mount block device 'ablkdev' successfully when HostPathType is HostPathUnset", func() {
+		verifyPodHostPathType(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName}, targetBlockDev, &hostPathUnset)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathDirectory", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a directory", targetBlockDev), &hostPathDirectory)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathFile", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a file", targetBlockDev), &hostPathFile)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathSocket", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a socket", targetBlockDev), &hostPathSocket)
+	})
+
+	ginkgo.It("Should fail on mounting block device 'ablkdev' when HostPathType is HostPathCharDev", func() {
+		verifyPodHostPathTypeFailure(f, map[string]string{"kubernetes.io/hostname": basePod.Spec.NodeName},
+			targetBlockDev, fmt.Sprintf("%s is not a character device", targetBlockDev), &hostPathCharDev)
+	})
+})
+
+func newHostPathTypeTestPod(nodeSelector map[string]string, hostDir, mountDir string, hostPathType *v1.HostPathType) *v1.Pod {
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hostpath-type-",
+		},
+		Spec: v1.PodSpec{
+			NodeSelector:  nodeSelector,
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:  "host-path-testing",
+					Image: imageutils.GetE2EImage(imageutils.Agnhost),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "host",
+							MountPath: mountDir,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+			Volumes: []v1.Volume{
+				{
+					Name: "host",
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: hostDir,
+							Type: hostPathType,
+						},
+					},
+				},
+			},
+		},
+	}
+	return pod
+}
+
+func verifyPodHostPathTypeFailure(f *framework.Framework, nodeSelector map[string]string, hostDir, pattern string, hostPathType *v1.HostPathType) {
+	pod := newHostPathTypeTestPod(nodeSelector, hostDir, "/mnt/test", hostPathType)
+	ginkgo.By(fmt.Sprintf("Creating pod %s", pod.Name))
+	pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(context.TODO(), pod, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+
+	err = e2epod.WaitTimeoutForPodRunningInNamespace(f.ClientSet, pod.Name, pod.Namespace, framework.PodStartShortTimeout)
```

Comment:
I think we can remove this check. The new flow is something like:

1. Create pod
2. If we get the event, then the test succeeds.
3. If we don't get the event after timeout, then check that the pod is still not running.

This can help speed up the test in the successful case, so we don't always have to for `PodStartShortTimeout`.